cpp/tests/inpainting.cc

```diff
@@ -0,0 +1,98 @@
+#include <algorithm>
+#include <exception>
+#include <functional>
+#include <iostream>
+#include <random>
+#include <vector>
+#include <ctime>
+#include <catch.hpp>
+
+#include <sopt/imaging_forward_backward.h>
+#include <sopt/logging.h>
+#include <sopt/maths.h>
+#include <sopt/relative_variation.h>
+#include <sopt/sampling.h>
+#include <sopt/types.h>
+#include <sopt/utilities.h>
+#include <sopt/wavelets.h>
+
+// This header is not part of the installed sopt interface
+// It is only present in tests
+#include <tools_for_tests/directories.h>
+#include <tools_for_tests/tiffwrappers.h>
+
+// \min_{x} ||\Psi^Tx||_1 \quad \mbox{s.t.} \quad ||y - Ax||_2 < \epsilon and x \geq 0
+
+typedef double Scalar;
+typedef sopt::Vector<Scalar> Vector;
+typedef sopt::Matrix<Scalar> Matrix;
+typedef sopt::Image<Scalar> Image;
+
+TEST_CASE("Inpainting"){
+  extern std::unique_ptr<std::mt19937_64> mersenne;
+  std::string const input = "cameraman256";
+  std::string const output = "none";
+
+  Image const image = sopt::notinstalled::read_standard_tiff(input);
+
+  sopt::t_uint nmeasure = std::floor(0.5 * image.size());
+  sopt::LinearTransform<Vector> const sampling =
+      sopt::linear_transform<Scalar>(sopt::Sampling(image.size(), nmeasure, *mersenne));
+
+  auto const wavelet = sopt::wavelets::factory("DB8", 4);
+
+  auto const psi = sopt::linear_transform<Scalar>(wavelet, image.rows(), image.cols());
+
+  Vector const y0 = sampling * Vector::Map(image.data(), image.size());
+  auto const snr = 30.0;
+  auto const sigma = y0.stableNorm() / std::sqrt(y0.size()) * std::pow(10.0, -(snr / 20.0));
+  auto const epsilon = std::sqrt(nmeasure + 2 * std::sqrt(y0.size())) * sigma;
+
+
+  std::normal_distribution<> gaussian_dist(0, sigma);
+  Vector y(y0.size());
+  for (sopt::t_int i = 0; i < y0.size(); i++) y(i) = y0(i) + gaussian_dist(*mersenne);
+  // Write dirty imagte to file
+  if (output != "none") {
```

Comment:
Probably don't want to be saving images out during unit testing. The probably wouldn't be called either way but imo better to just remove :)